### PR TITLE
8326135: Enhance adlc to report unused operands

### PR DESCRIPTION
**Pull request by kuaiwei**
Some operands are defined in adfile but no one used them. But it's hard to find them manually. So I try to enhance adlc to report them after parsing the whole adfile.

I added a helper Form::forms_do to recursively visit child forms. After parsing, adlc will start from all instructions to mark all used forms. And report unvisited operands as unused. By this way, I can find 44 unused operands for aarch64, 12 for x86_64 and 4 for riscv64. The report is like
```
...
Warning: unused operand (vRegD_V28)
Warning: unused operand (vRegD_V29)
Warning: unused operand (vRegD_V30)
Warning: unused operand (vRegD_V31)
Warning: unused operand (lr_RegP)
Warning: unused operand (indOffI)
Warning: unused operand (indOffL)
Warning: unused operand (thread_anchor_pc)
-------Warning: total 44 unused operands
```

I tested and find they can be safely removed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326135](https://bugs.openjdk.org/browse/JDK-8326135): Enhance adlc to report unused operands (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17910/head:pull/17910` \
`$ git checkout pull/17910`

Update a local copy of the PR: \
`$ git checkout pull/17910` \
`$ git pull https://git.openjdk.org/jdk.git pull/17910/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17910`

View PR using the GUI difftool: \
`$ git pr show -t 17910`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17910.diff">https://git.openjdk.org/jdk/pull/17910.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17910#issuecomment-1951972275)

### CODE REVIEW

File: src/hotspot/share/adlc/archDesc.cpp
Comment (vnkozlov) [inline src/hotspot/share/adlc/archDesc.cpp:739]: What is special about `stackSlot*` and `vec*` operands to process them explicitly here?

**Review by vnkozlov — COMMENTED** [review]
Very nice feature.

**Comment by vnkozlov** [comment]
What effect it has on `adlc` run time?

**Comment by kuaiwei** [comment]
> What effect it has on `adlc` run time?

It only reports warning. Not find impact on run time.

File: src/hotspot/share/adlc/archDesc.cpp
Comment (vnkozlov) [reply src/hotspot/share/adlc/archDesc.cpp:739]: I actually asked why `callback.do_form(instr)` calls in loop are not marking them as used? Why you need to call `do_form_by_name()` for them?

Comment (kuaiwei) [reply src/hotspot/share/adlc/archDesc.cpp:739]: For example, stackSlotP in aarch64 is not used. If not do_form_by_name, it will be reported as unused.

Comment (vnkozlov) [reply src/hotspot/share/adlc/archDesc.cpp:739]: Got it. Please, add comment describing this special case for vector and stack operators.

Comment (kuaiwei) [reply src/hotspot/share/adlc/archDesc.cpp:741]: Thanks for suggestion. I added the example in comments.

Comment (vnkozlov) [reply src/hotspot/share/adlc/archDesc.cpp:741]: Good. Please update comment for vectors too (add `only`):
```
// special generic vector operands only used in Matcher::pd_specialize_generic_vector_operand
```

I will submit our testing meanwhile.

**Comment by vnkozlov** [comment]
Please update Copyright year to 2024 in changed files if needed:
```
< // Copyright (c) 1997, 2023, Oracle and/or its affiliates. All rights reserved.
---
> // Copyright (c) 1997, 2024, Oracle and/or its affiliates. All rights reserved.
```
And you need second review

**Review by vnkozlov — APPROVED** [review]
My testing passed.

File: src/hotspot/share/adlc/archDesc.cpp
Comment (kuaiwei) [reply src/hotspot/share/adlc/archDesc.cpp]: comments fixed

**Comment by kuaiwei** [comment]
> Please update Copyright year to 2024 in changed files if needed:
> 
> ```
> < // Copyright (c) 1997, 2023, Oracle and/or its affiliates. All rights reserved.
> ---
> > // Copyright (c) 1997, 2024, Oracle and/or its affiliates. All rights reserved.
> ```
> 
> And you need second review

Copyright year are updated.

File: src/hotspot/share/adlc/archDesc.cpp
Comment (iwanowww) [inline src/hotspot/share/adlc/archDesc.cpp:27]: Even though it's not hotspot code per se (but a tool used during build), it'll be the first usage of C++ STL in the hotspot code base.

HotSpot Coding Style [1] clearly states "Avoid using the C++ Standard Library." 

@kimbarrett what are your thought on it?

[1] https://github.com/openjdk/jdk/blob/master/doc/hotspot-style.md

**Review by iwanowww — COMMENTED** [review]
Nice feature!

File: src/hotspot/share/adlc/archDesc.cpp
Comment (kuaiwei) [reply src/hotspot/share/adlc/archDesc.cpp:27]: I think adlc is almost a stand alone tool, it only depends on src/hotspot/share/opto/opcodes.cpp . Could we move it to src/utils?

**Comment by kuaiwei** [comment]
@vnkozlov I checked build of other architecture and find a false alert in arm. sReg* need to be special handle. Could you review the new patch?

File: src/hotspot/share/adlc/archDesc.cpp
Comment (kuaiwei) [reply src/hotspot/share/adlc/archDesc.cpp:27]: I'm thinking if it can be out of hotspot and use stl in code.

**Review by vnkozlov — APPROVED** [review]
Update is good.

File: src/hotspot/share/adlc/archDesc.cpp
Comment (vnkozlov) [reply src/hotspot/share/adlc/archDesc.cpp:27]: The rule is applied to general JDK sources and not just VM.
Please, check "C++ Standard Library" in `hotspot-style` Vladimir pointed and see if listed issues there applied to your changes.

Comment (kuaiwei) [reply src/hotspot/share/adlc/archDesc.cpp:27]: Ok, I will add a hashset implementation.

Comment (vnkozlov) [reply src/hotspot/share/adlc/archDesc.cpp:27]: Please wait. We are still discussing with other developers if it is OK to use STL in adlc.

Comment (vnkozlov) [reply src/hotspot/share/adlc/archDesc.cpp:27]: @kuaiwei  our general consensus is that you can use STL here since it is separate from VM tool.
@iwanowww do you approve?

Comment (kuaiwei) [reply src/hotspot/share/adlc/archDesc.cpp:27]: @iwanowww Can you check if I can use STL here?

Comment (iwanowww) [reply src/hotspot/share/adlc/archDesc.cpp:27]: Nobody spoke up against, so I'm fine with using it.

**Review by iwanowww — APPROVED** [review]

File: src/hotspot/share/adlc/archDesc.cpp
Comment (kuaiwei) [reply src/hotspot/share/adlc/archDesc.cpp:27]: Thanks

**Comment by kuaiwei** [comment]
@vnkozlov @iwanowww Can you help sponsor it? Thanks.

**Comment by vnkozlov** [comment]
@kuaiwei  I see

**Comment by kuaiwei** [comment]
> @kuaiwei I see new warnings when I build on AVX2 machine (no AVX512) with these changes. I filed [JDK-8326983](https://bugs.openjdk.org/browse/JDK-8326983). Please look.

Thanks for testing. I will check it.

**Comment by kuaiwei** [comment]
@vnkozlov My openjdk account has some problem and I can not update JBS issue. So I update my test result here.
In my testing on machine with AVX512, it report the same warning operands.
```
Warning: unused operand (immI_32)
Warning: unused operand (immI_64)
Warning: unused operand (immL10)
Warning: unused operand (immL_127)
Warning: unused operand (kReg_K1)
Warning: unused operand (kReg_K2)
Warning: unused operand (kReg_K3)
Warning: unused operand (kReg_K4)
Warning: unused operand (kReg_K5)
Warning: unused operand (kReg_K6)
Warning: unused operand (kReg_K7)
Warning: unused operand (no_rax_rbx_RegP)
-------Warning: total 12 unused operands
```
The adlc doesn't use machine specific feature to compile ad file. So it should has same behavior. There operands are real useless and can be cleaned. I will open a new issue to follow it. But I need fix my account issue first. :<

**Comment by vnkozlov** [comment]
> The adlc doesn't use machine specific feature to compile ad file. So it should has same behavior. There operands are real useless and can be cleaned. I will open a new issue to follow it. But I need fix my account issue first. :<

Yes, I observed it on AVX512 machine as well. Please, don't open new issue - use [JDK-8326983](https://bugs.openjdk.org/browse/JDK-8326983) for it. I made it general to fix issue on x86 and aarch64.

**Pull request closed without merging**
